<a href="https://colab.research.google.com/github/Shopping-Yuan/ML2021HW/blob/Shopping_vscode_branch/HW4/HW04_modified.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import shutil
shutil.copyfile('/content/drive/MyDrive/Dataset.zip','/content/Dataset.zip')

import zipfile
with zipfile.ZipFile('/content/Dataset.zip','r') as zip_ref:
  zip_ref.extractall('/content/')

In [56]:
def parse_args():
  """arguments"""
  config = {
    "data_dir": "/content/Dataset",
    "save_path": "model.ckpt",
    "batch_size": 32,
    "n_workers": 8,
    "valid_steps": 2000,
    "warmup_steps": 1000,
    "save_steps": 10000,
    "total_steps": 70000,
  }

  return config

#Dataset

In [57]:
import os
import json
import torch
import random
from pathlib import Path
from torch.utils.data import Dataset
from torch.nn.utils.rnn import pad_sequence


class myDataset(Dataset):
  def __init__(self, data_dir, segment_len=128):
    self.data_dir = data_dir
    self.segment_len = segment_len

    # Load the mapping from speaker neme to their corresponding id.
    # Path package can use / to join path
    #part 1-1
    mapping_path = Path(data_dir) / "mapping.json"
    mapping = json.load(mapping_path.open())
    self.speaker2id = mapping["speaker2id"]
    #self.speaker2id = {'id10001': 0, 'id10005': 1, ...... 'id11247': 598, 'id11251': 599}

    #part 1-2
    # Load metadata of training data.
    metadata_path = Path(data_dir) / "metadata.json"
    metadata = json.load(open(metadata_path))["speakers"]

    # Get the total number of speaker.
    self.speaker_num = len(metadata.keys())
    self.data = []
    for speaker in metadata.keys():
    #['id10473', 'id10328', 'id10382',......'id10864']
      for utterances in metadata[speaker]:
      # for example : metadata[speaker] = 'id10473'
      # utterances = {'feature_path': 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt', 'mel_len': 652}......
        self.data.append([utterances["feature_path"], self.speaker2id[speaker]])
        # for example : ['uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt',217]

  def __len__(self):
    return len(self.data)

  def __getitem__(self, index):
    feat_path, speaker = self.data[index]
    #[utterances["feature_path"], self.speaker2id[speaker]

    # Load preprocessed mel-spectrogram.
    # https://pytorch.org/docs/stable/generated/torch.load.html
    # using torch.load to open pt file . By default , tensors are loaded to GPU if they are GPU tensor.
    # part 2
    mel = torch.load(os.path.join(self.data_dir, feat_path))

    # Segmemt mel-spectrogram into "segment_len" frames.
    if len(mel) > self.segment_len:
      # Randomly get the starting point of the segment.
      start = random.randint(0, len(mel) - self.segment_len)
      # Get a segment with "segment_len" frames.
      mel = torch.FloatTensor(mel[start:start+self.segment_len])
    else:
      mel = torch.FloatTensor(mel)
    # Turn the speaker id into long for computing loss later.
    speaker = torch.FloatTensor([speaker]).long()
    return mel, speaker

  def get_speaker_number(self):
    return self.speaker_num

In [ ]:
#data_dir = parse_args()["data_dir"]
#part 1-1
#mapping.json checcking
#mapping_path = Path(data_dir) / "mapping.json"
#mapping = json.load(mapping_path.open())
#speaker2id = mapping["speaker2id"]
#print(type(mapping)) #<class 'dict'>
#print(mapping.keys()) #dict_keys(['speaker2id', 'id2speaker'])
#print(mapping.values()) #dict_values([{'id10001': 0, 'id10005': 1, ...... 'id11247': 598, 'id11251': 599},
             #{"0": "id10001", ......'599':  'id11251'}])
#print(mapping['speaker2id']['id10473']) #217

#part 1-2
#metadata.json checcking
#metadata_path = Path(data_dir) / "metadata.json"
#metadata = json.load(open(metadata_path))["speakers"]
#print(type(metadata)) #<class 'dict'>
#print(len(metadata.keys()),metadata.keys()) #600 dict_keys(['id10473', 'id10328', 'id10382',......'id10864'])
#print(len(metadata.values()),len(list(metadata.values())[0]),list(metadata.values())[0])
#600 176 [{'feature_path': 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt', 'mel_len': 652},......
#{'feature_path': 'uttr-caf385f0032140e0add0bbde9a6595cd.pt', 'mel_len': 414}]
#print(len(metadata.values()),len(list(metadata.values())[1])) #600 101

#part 2
#mel = torch.load(os.path.join(data_dir, 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt'))
#print(type(mel)) #<class 'torch.Tensor'>
#print(mel.size()) #torch.Size([652, 40])
#print(mel[0])
# tensor([
# 2.0060,  1.9015,  1.3473,  1.5711,  1.5569,  2.3200,  3.0064,  3.3848,
# 0.5376,  0.8397,  1.5710, -0.5662, -2.1262, -1.1271, -1.4272, -3.6233,
# -1.0765, -1.8453, -3.2415, -2.3427, -1.5888, -0.4692, -0.4291, -0.4722,
# -1.8450, -1.0638, -3.2177, -3.4006, -3.2176, -3.3409, -3.1142, -4.0395,
# -3.6502, -4.2407, -4.9656, -5.1238, -5.0620, -5.4308, -6.0757, -6.6186])

In [58]:
import torch
from torch.utils.data import DataLoader, random_split
from torch.nn.utils.rnn import pad_sequence
#part 1 check collate_batch
def collate_batch(batch):
  # Process features within a batch.
  """Collate a batch of data."""
  #part 1-1 , 1-2
  mel, speaker = zip(*batch)
  # Because we train the model batch by batch, we need to pad the features in the same batch to make their lengths the same.
  mel = pad_sequence(mel, batch_first=True, padding_value=-20)    # pad log 10^(-20) which is very small value.
  # mel: (batch size, length, 40)
  #part 1-3
  return mel, torch.FloatTensor(speaker).long()


def get_dataloader(data_dir, batch_size, n_workers):
  """Generate dataloader"""
  dataset = myDataset(data_dir)
  speaker_num = dataset.get_speaker_number()
  # Split dataset into training dataset and validation dataset
  trainlen = int(0.9 * len(dataset))
  lengths = [trainlen, len(dataset) - trainlen]
  trainset, validset = random_split(dataset, lengths)

  train_loader = DataLoader(
    trainset,
    batch_size=batch_size,
    shuffle=True,
    drop_last=True,
    num_workers=n_workers,
    pin_memory=True,
    collate_fn=collate_batch,
  )
  valid_loader = DataLoader(
    validset,
    batch_size=batch_size,
    num_workers=n_workers,
    drop_last=True,
    pin_memory=True,
    collate_fn=collate_batch,
  )

  return train_loader, valid_loader, speaker_num

In [ ]:
# x = [["a",1],["b",2],["c",3]]
# print(list(zip(*x)))
# print(list(zip(("a",1),("b",2),("c",3))))

In [ ]:
#import numpy as np
#check collate_batch
#part 1-1
# * collect "sequence"function input to a tuple :
#def f(*paras):
#  print(paras)
#f(1,2,3) #(1, 2, 3)
# or collect a tuple to a "sequence"
#print(*(1,2,3)) #1 2 3

#part 1-2
#test_data = np.random.rand(2,3,4)
#test_label = np.random.rand(2,1)
#samples = (test_data,test_label)
#print(samples)
#x , y = zip((1,2),(3,4),(5,6),(7,8))
#print(x,y)

#part 1-3
#print(torch.FloatTensor([[1.],[1.]]))#tensor([[1.],[1.]])
#print(torch.FloatTensor([[1.],[1.]]).long())#tensor([[1],[1]]) implies the numbers are integer now
#print(torch.IntTensor([[1.],[1.]]))#tensor([[1],[1]], dtype=torch.int32)

In [ ]:
!pip install torchinfo

In [ ]:
# x = [True,True,False,True,False,False,False,True,False,True,True,False]

# true_index_list = []

# false_index_list = []

# for i,b in enumerate(x):
#   if b == True:
#     true_index_list += [i]
#   else :
#     false_index_list += [i]
# print(true_index_list,false_index_list)

# new_list = []
# t_max = len(true_index_list)
# f_max = len(false_index_list)
# true_index = 0
# false_index = 0
# t_or_f = 0
# while (t_max * f_max)>0:
#   if t_or_f % 2 ==0:
#     new_list += true_index_list[true_index:true_index+3]
#     true_index += 3
#     t_or_f += 1
#     t_max -= 3
#   else :
#     new_list += false_index_list[false_index:false_index+3]
#     false_index += 3
#     t_or_f += 1
#     f_max -= 3
# if t_max > 0:
#   new_list += true_index_list[true_index:]
# elif f_max > 0:
#   new_list += false_index_list[false_index:]

# print(new_list)


In [59]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
from torchinfo import summary

class Classifier(nn.Module):
  def __init__(self, d_model=128, n_spks=600, dropout=0.1):
    super().__init__()
    # Project the dimension of features from that of input into d_model.
    self.prenet = nn.Linear(40, d_model)
    # TODO:
    #   Change Transformer to Conformer.
    #   https://arxiv.org/abs/2005.08100
    # embed_dim must be divisible by num_heads
    self.encoder_layer = nn.TransformerEncoderLayer(
      d_model=d_model, dim_feedforward=256, nhead=1
    )
    # batch_first = True
    # norm_first = True
    self.encoder = nn.TransformerEncoder(self.encoder_layer, num_layers=4)

    # Project the the dimension of features from d_model into speaker nums.
    self.pred_layer = nn.Sequential(
      nn.Linear(d_model, d_model),
      nn.ReLU(),
      nn.Linear(d_model, n_spks),
    )

  def forward(self, mels):
    """
    args:
      mels: (batch size, length, 40)
    return:
      out: (batch size, n_spks)
    """
    # out: (batch size, length, d_model)
    out = self.prenet(mels)
    # out: (length, batch size, d_model)
    out = out.permute(1, 0, 2)
    # The encoder layer expect features in the shape of (length, batch size, d_model).
    #out = self.encoder_layer(out)
    out = self.encoder(out)
    # out: (batch size, length, d_model)
    out = out.transpose(0, 1)
    # mean pooling
    stats = out.mean(dim=1)

    # out: (batch, n_spks)
    out = self.pred_layer(stats)
    return out


In [60]:
model = Classifier()
print(summary(model,(32,128,40)))
#print(model.state_dict().keys())
#print(model.state_dict()["encoder_layer.self_attn.in_proj_weight"].size())#torch.Size([240, 80]) (80+80+80)*80
#print(model.state_dict()["encoder_layer.self_attn.out_proj.weight"].size())#torch.Size([80, 80])

Layer (type:depth-idx)                   Output Shape              Param #
Classifier                               [32, 600]                 529,920
├─Linear: 1-1                            [32, 128, 128]            5,248
├─TransformerEncoderLayer: 1-2           [128, 32, 128]            --
│    └─MultiheadAttention: 2-1           [128, 32, 128]            66,048
│    └─Dropout: 2-2                      [128, 32, 128]            --
│    └─LayerNorm: 2-3                    [128, 32, 128]            256
│    └─Linear: 2-4                       [128, 32, 256]            33,024
│    └─Dropout: 2-5                      [128, 32, 256]            --
│    └─Linear: 2-6                       [128, 32, 128]            32,896
│    └─Dropout: 2-7                      [128, 32, 128]            --
│    └─LayerNorm: 2-8                    [128, 32, 128]            256
├─Sequential: 1-3                        [32, 600]                 --
│    └─Linear: 2-9                       [32, 128]             

/usr/local/lib/python3.10/dist-packages/torch/nn/modules/transformer.py:282: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.batch_first was not True(use batch_first for better inference performance)
  warnings.warn(f"enable_nested_tensor is True, but self.use_nested_tensor is False because {why_not_sparsity_fast_path}")


original
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (80+1)80 + (80+1)80 + (80+1)80 + (80+1)80 = 25920
2.layer_norm : =input + input = 80*80 = 160

first change
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (128+1)128(3+1)= 16512

second change
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (128+1)128(3+1)= 16512
two layers : 33024

2.layer_norm : =input + input = 80*80 = 160

3rd change
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (128+1)128(3+1)= 16512
four layers : 66048

2.layer_norm : =input + input = 80*80 = 160


In [61]:
import math

import torch
from torch.optim import Optimizer
from torch.optim.lr_scheduler import LambdaLR


def get_cosine_schedule_with_warmup(
  optimizer: Optimizer,
  num_warmup_steps: int,
  num_training_steps: int,
  num_cycles: float = 0.5,
  last_epoch: int = -1,
):
  """
  Create a schedule with a learning rate that decreases following the values of the cosine function between the
  initial lr set in the optimizer to 0, after a warmup period during which it increases linearly between 0 and the
  initial lr set in the optimizer.

  Args:
    optimizer (:class:`~torch.optim.Optimizer`):
      The optimizer for which to schedule the learning rate.
    num_warmup_steps (:obj:`int`):
      The number of steps for the warmup phase.
    num_training_steps (:obj:`int`):
      The total number of training steps.
    num_cycles (:obj:`float`, `optional`, defaults to 0.5):
      The number of waves in the cosine schedule (the defaults is to just decrease from the max value to 0
      following a half-cosine).
    last_epoch (:obj:`int`, `optional`, defaults to -1):
      The index of the last epoch when resuming training.

  Return:
    :obj:`torch.optim.lr_scheduler.LambdaLR` with the appropriate schedule.
  """

  def lr_lambda(current_step):
    # Warmup
    if current_step < num_warmup_steps:
      return float(current_step) / float(max(1, num_warmup_steps))
    # decadence
    # cosine angle
    progress = float(current_step - num_warmup_steps) / float(
      max(1, num_training_steps - num_warmup_steps)
    )
    return max(
      0.0, 0.5 * (1.0 + math.cos(math.pi * float(num_cycles) * 2.0 * progress))
    )

  return LambdaLR(optimizer, lr_lambda, last_epoch)

In [62]:
from torch.optim import AdamW
optimizer = AdamW(model.parameters(), lr=1e-3)

scheduler = get_cosine_schedule_with_warmup(optimizer, 1000, 70000)

#print(type(scheduler.optimizer.param_groups)) #<class 'list'>
#print(len(scheduler.optimizer.param_groups))  #1
#print(type(scheduler.optimizer.param_groups[0])) #<class 'dict'>
#print(scheduler.optimizer.param_groups[0].keys())
#['params', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad', 'foreach',
#'maximize', 'capturable', 'differentiable', 'fused', 'initial_lr']
#print(scheduler.optimizer.param_groups[0]['params'])
#[Parameter containing: tensor([[......],requires_grad=True)]
#print(scheduler.optimizer.param_groups[0]['lr'])# 0.0
#print(scheduler.optimizer.param_groups[0]['betas'])# (0.9, 0.999)
#print(scheduler.optimizer.param_groups[0]['eps']) # 1e-08
#print(scheduler.optimizer.param_groups[0]['weight_decay']) #0.01
#print(scheduler.optimizer.param_groups[0]['amsgrad']) # False
#print(scheduler.optimizer.param_groups[0]['foreach']) # None

In [63]:
import torch


def model_fn(batch, model, criterion, device):
  """Forward a batch through the model."""

  mels, labels = batch
  mels = mels.to(device)
  labels = labels.to(device)

  outs = model(mels)

  loss = criterion(outs, labels)

  # Get the speaker id with highest probability.
  preds = outs.argmax(1)
  # Compute accuracy.
  accuracy = torch.mean((preds == labels).float())

  return loss, accuracy

In [64]:
from tqdm import tqdm
import torch


def valid(dataloader, model, criterion, device):
  """Validate on validation set."""

  model.eval()
  running_loss = 0.0
  running_accuracy = 0.0
  # Manually update the progress bar : step 1
  # total length
  pbar = tqdm(total=len(dataloader.dataset), ncols=0, desc="Valid", unit=" uttr")
  for i, batch in enumerate(dataloader):
    with torch.no_grad():
      loss, accuracy = model_fn(batch, model, criterion, device)
      running_loss += loss.item()
      running_accuracy += accuracy.item()
    # Manually update the progress bar : step 2
    # update each batch
    pbar.update(dataloader.batch_size)
    pbar.set_postfix(
      loss=f"{running_loss / (i+1):.2f}",
      accuracy=f"{running_accuracy / (i+1):.2f}",
    )
  # Manually update the progress bar : step 3
  pbar.close()
  model.train()

  return running_accuracy / len(dataloader)


In [65]:
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.optim import AdamW
from torch.utils.data import DataLoader, random_split


# def parse_args():
#   """arguments"""
#   config = {
#     "data_dir": "/content/Dataset",
#     "save_path": "model.ckpt",
#     "batch_size": 32,
#     "n_workers": 8,
#     "valid_steps": 2000,
#     "warmup_steps": 1000,
#     "save_steps": 10000,
#     "total_steps": 70000,
#   }

#   return config


def main(
  data_dir,
  save_path,
  batch_size,
  n_workers,
  valid_steps,
  warmup_steps,
  total_steps,
  save_steps,
):
  """Main function."""
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  print(f"[Info]: Use {device} now!")

  train_loader, valid_loader, speaker_num = get_dataloader(data_dir, batch_size, n_workers)
  train_iterator = iter(train_loader)
  print(f"[Info]: Finish loading data!",flush = True)

  model = Classifier(n_spks=speaker_num).to(device)
  criterion = nn.CrossEntropyLoss()
  optimizer = AdamW(model.parameters(), lr=1e-3)
  scheduler = get_cosine_schedule_with_warmup(optimizer, warmup_steps, total_steps)
  print(f"[Info]: Finish creating model!",flush = True)

  best_accuracy = -1.0
  best_state_dict = None
  # pbar has total 2000 step
  pbar = tqdm(total=valid_steps, ncols=0, desc="Train", unit=" step")
  # print train loss and accuracy every 1000 step

  loss_acc_list = []

  for step in range(total_steps):
    # Get data
    try:
      batch = next(train_iterator)
    except StopIteration:
      train_iterator = iter(train_loader)
      batch = next(train_iterator)

    loss, accuracy = model_fn(batch, model, criterion, device)
    batch_loss = loss.item()
    batch_accuracy = accuracy.item()

    # Updata model
    loss.backward()
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
    # print train loss acc
    loss_acc_list.append((loss, accuracy))
    print_loss , print_acc = list(zip(*loss_acc_list))
    if step % 1000 == 0:
      print(f"average loss = {sum(print_loss)/len(print_loss):.4f} \
      average accuracy = {sum(print_acc)/len(print_acc):.4f})")
      loss_acc_list = []

    # Log
    # update each step (as total)
    pbar.update()
    pbar.set_postfix(
      loss=f"{batch_loss:.2f}",
      accuracy=f"{batch_accuracy:.2f}",
      step=step + 1,
    )

    # Do validation
    # valid each 2000 step
    if (step + 1) % valid_steps == 0:
      #close above pbar
      pbar.close()

      valid_accuracy = valid(valid_loader, model, criterion, device)

      # keep the best model
      if valid_accuracy > best_accuracy:
        best_accuracy = valid_accuracy
        best_state_dict = model.state_dict()
      # new pbar has total 2000 step
      pbar = tqdm(total=valid_steps, ncols=0, desc="Train", unit=" step")

    # Save the best model so far.
    if (step + 1) % save_steps == 0 and best_state_dict is not None:
      torch.save(best_state_dict, save_path)
      pbar.write(f"Step {step + 1}, best model saved. (accuracy={best_accuracy:.4f})")

  pbar.close()


if __name__ == "__main__":
  main(**parse_args())

[Info]: Use cuda now!


/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:557: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


[Info]: Finish loading data!
[Info]: Finish creating model!


Train:   0% 1/2000 [00:00<05:07,  6.51 step/s, accuracy=0.00, loss=6.45, step=1]

average loss = 6.4498       average accuracy = 0.0000)


Train:  50% 1001/2000 [00:52<01:00, 16.45 step/s, accuracy=0.09, loss=4.54, step=1001]

average loss = 5.6269       average accuracy = 0.0454)


Train: 100% 2000/2000 [01:58<00:00, 16.84 step/s, accuracy=0.25, loss=3.29, step=2000]
Valid: 100% 6944/6944 [00:15<00:00, 455.08 uttr/s, accuracy=0.22, loss=3.72]
Train:   0% 1/2000 [00:00<03:48,  8.74 step/s, accuracy=0.19, loss=3.78, step=2001]

average loss = 4.0981       average accuracy = 0.1763)


Train:  50% 1007/2000 [00:56<00:55, 17.89 step/s, accuracy=0.22, loss=4.08, step=3007]

average loss = 3.4555       average accuracy = 0.2675)


Train: 100% 2000/2000 [01:56<00:00, 17.16 step/s, accuracy=0.38, loss=2.91, step=4000]
Valid: 100% 6944/6944 [00:14<00:00, 473.77 uttr/s, accuracy=0.33, loss=3.05]
Train:   0% 9/2000 [00:00<00:46, 43.06 step/s, accuracy=0.56, loss=2.12, step=4009]

average loss = 3.1698       average accuracy = 0.3157)


Train:  50% 1008/2000 [00:55<01:10, 13.98 step/s, accuracy=0.41, loss=2.81, step=5008]

average loss = 2.9090       average accuracy = 0.3608)


Train: 100% 2000/2000 [01:55<00:00, 17.31 step/s, accuracy=0.47, loss=2.86, step=6000]
Valid: 100% 6944/6944 [00:14<00:00, 487.94 uttr/s, accuracy=0.39, loss=2.74]
Train:   0% 5/2000 [00:00<03:11, 10.43 step/s, accuracy=0.34, loss=2.73, step=6005]

average loss = 2.7632       average accuracy = 0.3860)


Train:  50% 1004/2000 [00:55<01:11, 13.85 step/s, accuracy=0.44, loss=2.20, step=7004]

average loss = 2.6008       average accuracy = 0.4107)


Train: 100% 2000/2000 [01:52<00:00, 17.73 step/s, accuracy=0.44, loss=3.14, step=8000]
Valid: 100% 6944/6944 [00:14<00:00, 490.28 uttr/s, accuracy=0.44, loss=2.49]
Train:   0% 7/2000 [00:00<02:46, 11.98 step/s, accuracy=0.53, loss=2.26, step=8007]

average loss = 2.5204       average accuracy = 0.4270)


Train:  50% 1005/2000 [00:56<01:10, 14.02 step/s, accuracy=0.31, loss=3.08, step=9005]

average loss = 2.4261       average accuracy = 0.4473)


Train: 100% 2000/2000 [01:53<00:00, 17.58 step/s, accuracy=0.41, loss=2.32, step=1e+4]
Valid: 100% 6944/6944 [00:14<00:00, 475.52 uttr/s, accuracy=0.46, loss=2.40]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 5/2000 [00:00<02:17, 14.47 step/s, accuracy=0.50, loss=2.62, step=1e+4]

Step 10000, best model saved. (accuracy=0.4595)
average loss = 2.3199       average accuracy = 0.4682)


Train:  50% 1008/2000 [00:59<01:07, 14.77 step/s, accuracy=0.50, loss=2.79, step=11008]

average loss = 2.2256       average accuracy = 0.4860)


Train: 100% 2000/2000 [01:55<00:00, 17.28 step/s, accuracy=0.47, loss=2.11, step=12000]
Valid: 100% 6944/6944 [00:13<00:00, 497.09 uttr/s, accuracy=0.48, loss=2.31]
Train:   0% 9/2000 [00:00<02:15, 14.67 step/s, accuracy=0.59, loss=1.96, step=12009]

average loss = 2.1719       average accuracy = 0.4949)


Train:  50% 1003/2000 [01:01<01:13, 13.65 step/s, accuracy=0.53, loss=2.15, step=13003]

average loss = 2.1120       average accuracy = 0.5082)


Train: 100% 2000/2000 [01:57<00:00, 17.03 step/s, accuracy=0.50, loss=2.28, step=14000]
Valid: 100% 6944/6944 [00:14<00:00, 481.70 uttr/s, accuracy=0.51, loss=2.16]
Train:   0% 5/2000 [00:00<03:30,  9.47 step/s, accuracy=0.59, loss=1.87, step=14005]

average loss = 2.0557       average accuracy = 0.5131)


Train:  50% 1003/2000 [01:01<01:01, 16.16 step/s, accuracy=0.56, loss=1.56, step=15003]

average loss = 1.9810       average accuracy = 0.5304)


Train: 100% 2000/2000 [01:55<00:00, 17.27 step/s, accuracy=0.66, loss=1.41, step=16000]
Valid: 100% 6944/6944 [00:14<00:00, 468.83 uttr/s, accuracy=0.53, loss=2.07]
Train:   0% 10/2000 [00:00<02:31, 13.15 step/s, accuracy=0.75, loss=1.24, step=16010]

average loss = 1.9156       average accuracy = 0.5459)


Train:  50% 1005/2000 [01:00<00:57, 17.20 step/s, accuracy=0.62, loss=1.86, step=17005]

average loss = 1.8887       average accuracy = 0.5515)


Train: 100% 2000/2000 [01:52<00:00, 17.82 step/s, accuracy=0.53, loss=1.85, step=18000]
Valid: 100% 6944/6944 [00:14<00:00, 488.62 uttr/s, accuracy=0.53, loss=2.06]
Train:   0% 7/2000 [00:00<03:04, 10.79 step/s, accuracy=0.50, loss=1.93, step=18007]

average loss = 1.8167       average accuracy = 0.5677)


Train:  50% 1002/2000 [01:00<00:55, 18.11 step/s, accuracy=0.62, loss=1.70, step=19002]

average loss = 1.8057       average accuracy = 0.5652)


Train: 100% 2000/2000 [01:51<00:00, 18.01 step/s, accuracy=0.59, loss=1.62, step=2e+4]
Valid: 100% 6944/6944 [00:15<00:00, 453.32 uttr/s, accuracy=0.56, loss=1.93]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 1/2000 [00:00<04:49,  6.91 step/s, accuracy=0.44, loss=1.84, step=2e+4]

Step 20000, best model saved. (accuracy=0.5592)
average loss = 1.7471       average accuracy = 0.5815)


Train:  50% 1008/2000 [01:02<00:57, 17.18 step/s, accuracy=0.62, loss=1.95, step=21008]

average loss = 1.7014       average accuracy = 0.5899)


Train: 100% 2000/2000 [01:53<00:00, 17.55 step/s, accuracy=0.75, loss=1.11, step=22000]
Valid: 100% 6944/6944 [00:15<00:00, 460.19 uttr/s, accuracy=0.57, loss=1.88]
Train:   0% 8/2000 [00:00<02:35, 12.84 step/s, accuracy=0.72, loss=1.17, step=22008]

average loss = 1.6630       average accuracy = 0.5969)


Train:  50% 1008/2000 [01:04<01:14, 13.28 step/s, accuracy=0.50, loss=2.17, step=23008]

average loss = 1.6413       average accuracy = 0.6043)


Train: 100% 2000/2000 [01:55<00:00, 17.28 step/s, accuracy=0.66, loss=1.46, step=24000]
Valid: 100% 6944/6944 [00:15<00:00, 451.96 uttr/s, accuracy=0.59, loss=1.78]
Train:   1% 12/2000 [00:00<00:38, 51.44 step/s, accuracy=0.72, loss=1.44, step=24011]

average loss = 1.5882       average accuracy = 0.6156)


Train:  50% 1001/2000 [01:04<00:49, 20.01 step/s, accuracy=0.62, loss=1.61, step=25001]

average loss = 1.5916       average accuracy = 0.6144)


Train: 100% 2000/2000 [01:53<00:00, 17.61 step/s, accuracy=0.78, loss=1.02, step=26000]
Valid: 100% 6944/6944 [00:17<00:00, 390.21 uttr/s, accuracy=0.59, loss=1.78]
Train:   0% 9/2000 [00:00<00:53, 37.25 step/s, accuracy=0.62, loss=1.55, step=26009]

average loss = 1.5069       average accuracy = 0.6336)


Train:  50% 1006/2000 [01:03<01:08, 14.59 step/s, accuracy=0.66, loss=1.50, step=27006]

average loss = 1.5312       average accuracy = 0.6262)


Train: 100% 2000/2000 [01:54<00:00, 17.47 step/s, accuracy=0.66, loss=1.60, step=28000]
Valid: 100% 6944/6944 [00:14<00:00, 470.70 uttr/s, accuracy=0.60, loss=1.75]
Train:   0% 9/2000 [00:00<00:51, 38.81 step/s, accuracy=0.69, loss=1.16, step=28009]

average loss = 1.4828       average accuracy = 0.6374)


Train:  50% 1008/2000 [01:04<01:06, 14.92 step/s, accuracy=0.62, loss=1.88, step=29008]

average loss = 1.4378       average accuracy = 0.6463)


Train: 100% 2000/2000 [01:53<00:00, 17.68 step/s, accuracy=0.72, loss=1.43, step=3e+4]
Valid: 100% 6944/6944 [00:15<00:00, 458.09 uttr/s, accuracy=0.61, loss=1.68]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 6/2000 [00:00<02:08, 15.57 step/s, accuracy=0.62, loss=1.93, step=3e+4]

Step 30000, best model saved. (accuracy=0.6058)
average loss = 1.4192       average accuracy = 0.6507)


Train:  50% 1002/2000 [01:02<00:52, 19.14 step/s, accuracy=0.59, loss=1.77, step=31002]

average loss = 1.4047       average accuracy = 0.6541)


Train: 100% 2000/2000 [01:50<00:00, 18.06 step/s, accuracy=0.56, loss=1.49, step=32000]
Valid: 100% 6944/6944 [00:15<00:00, 455.57 uttr/s, accuracy=0.62, loss=1.64]
Train:   0% 7/2000 [00:00<02:37, 12.62 step/s, accuracy=0.81, loss=0.68, step=32007]

average loss = 1.3357       average accuracy = 0.6697)


Train:  50% 1002/2000 [01:04<00:59, 16.71 step/s, accuracy=0.69, loss=1.14, step=33002]

average loss = 1.3550       average accuracy = 0.6675)


Train: 100% 2000/2000 [01:54<00:00, 17.51 step/s, accuracy=0.59, loss=1.55, step=34000]
Valid: 100% 6944/6944 [00:15<00:00, 450.98 uttr/s, accuracy=0.63, loss=1.62]
Train:   0% 4/2000 [00:00<03:07, 10.66 step/s, accuracy=0.59, loss=1.51, step=34004]

average loss = 1.3039       average accuracy = 0.6784)


Train:  50% 1004/2000 [01:03<00:59, 16.73 step/s, accuracy=0.66, loss=1.69, step=35004]

average loss = 1.3212       average accuracy = 0.6738)


Train: 100% 2000/2000 [01:53<00:00, 17.67 step/s, accuracy=0.66, loss=1.42, step=36000]
Valid: 100% 6944/6944 [00:15<00:00, 435.62 uttr/s, accuracy=0.64, loss=1.57]
Train:   0% 10/2000 [00:00<00:43, 46.04 step/s, accuracy=0.78, loss=1.01, step=36010]

average loss = 1.2501       average accuracy = 0.6889)


Train:  50% 1003/2000 [01:03<01:03, 15.70 step/s, accuracy=0.62, loss=1.42, step=37003]

average loss = 1.2714       average accuracy = 0.6838)


Train: 100% 2000/2000 [01:51<00:00, 17.89 step/s, accuracy=0.56, loss=1.66, step=38000]
Valid: 100% 6944/6944 [00:15<00:00, 447.24 uttr/s, accuracy=0.65, loss=1.53]
Train:   1% 11/2000 [00:00<00:36, 53.78 step/s, accuracy=0.66, loss=1.39, step=38011]

average loss = 1.1913       average accuracy = 0.7020)


Train:  50% 1002/2000 [01:02<00:58, 17.19 step/s, accuracy=0.69, loss=1.23, step=39002]

average loss = 1.2282       average accuracy = 0.6950)


Train: 100% 2000/2000 [01:49<00:00, 18.34 step/s, accuracy=0.78, loss=1.14, step=4e+4]
Valid: 100% 6944/6944 [00:15<00:00, 440.78 uttr/s, accuracy=0.66, loss=1.46]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 5/2000 [00:00<02:18, 14.39 step/s, accuracy=0.78, loss=0.84, step=4e+4]

Step 40000, best model saved. (accuracy=0.6643)
average loss = 1.1463       average accuracy = 0.7134)


Train:  50% 1002/2000 [01:05<01:25, 11.62 step/s, accuracy=0.69, loss=1.09, step=41002]

average loss = 1.1699       average accuracy = 0.7110)


Train: 100% 2000/2000 [01:50<00:00, 18.03 step/s, accuracy=0.66, loss=1.34, step=42000]
Valid: 100% 6944/6944 [00:15<00:00, 455.06 uttr/s, accuracy=0.66, loss=1.48]
Train:   0% 10/2000 [00:00<00:40, 48.69 step/s, accuracy=0.66, loss=1.07, step=42010]

average loss = 1.1381       average accuracy = 0.7143)


Train:  50% 1004/2000 [01:03<01:01, 16.08 step/s, accuracy=0.75, loss=1.18, step=43004]

average loss = 1.1091       average accuracy = 0.7219)


Train: 100% 2000/2000 [01:52<00:00, 17.83 step/s, accuracy=0.78, loss=1.03, step=44000]
Valid: 100% 6944/6944 [00:15<00:00, 439.25 uttr/s, accuracy=0.68, loss=1.44]
Train:   0% 6/2000 [00:00<03:06, 10.67 step/s, accuracy=0.62, loss=1.57, step=44006]

average loss = 1.0748       average accuracy = 0.7318)


Train:  50% 1002/2000 [01:01<00:33, 30.22 step/s, accuracy=0.81, loss=0.84, step=45002]

average loss = 1.0854       average accuracy = 0.7284)


Train: 100% 2000/2000 [01:51<00:00, 18.00 step/s, accuracy=0.66, loss=1.01, step=46000]
Valid: 100% 6944/6944 [00:14<00:00, 467.75 uttr/s, accuracy=0.68, loss=1.38]
Train:   1% 13/2000 [00:00<00:35, 56.11 step/s, accuracy=0.78, loss=1.06, step=46013]

average loss = 1.0715       average accuracy = 0.7307)


Train:  50% 1003/2000 [01:00<00:40, 24.48 step/s, accuracy=0.66, loss=1.40, step=47003]

average loss = 1.0552       average accuracy = 0.7344)


Train: 100% 2000/2000 [01:51<00:00, 17.96 step/s, accuracy=0.81, loss=0.85, step=48000]
Valid: 100% 6944/6944 [00:15<00:00, 447.18 uttr/s, accuracy=0.69, loss=1.37]
Train:   0% 10/2000 [00:00<00:46, 43.11 step/s, accuracy=0.84, loss=0.68, step=48010]

average loss = 1.0217       average accuracy = 0.7450)


Train:  50% 1003/2000 [00:57<00:47, 20.80 step/s, accuracy=0.84, loss=0.52, step=49003]

average loss = 1.0258       average accuracy = 0.7442)


Train: 100% 2000/2000 [01:49<00:00, 18.26 step/s, accuracy=0.72, loss=1.14, step=5e+4]
Valid: 100% 6944/6944 [00:15<00:00, 439.17 uttr/s, accuracy=0.69, loss=1.36]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 2/2000 [00:00<04:15,  7.82 step/s, accuracy=0.66, loss=1.20, step=5e+4]

Step 50000, best model saved. (accuracy=0.6912)
average loss = 0.9743       average accuracy = 0.7558)


Train:  50% 1003/2000 [00:57<00:56, 17.66 step/s, accuracy=0.81, loss=0.69, step=51002]

average loss = 0.9962       average accuracy = 0.7509)


Train: 100% 2000/2000 [01:48<00:00, 18.35 step/s, accuracy=0.81, loss=0.70, step=52000]
Valid: 100% 6944/6944 [00:14<00:00, 469.15 uttr/s, accuracy=0.69, loss=1.34]
Train:   1% 11/2000 [00:00<00:39, 50.02 step/s, accuracy=0.81, loss=1.01, step=52011]

average loss = 0.9658       average accuracy = 0.7573)


Train:  50% 1001/2000 [00:57<00:39, 25.51 step/s, accuracy=0.69, loss=1.16, step=53001]

average loss = 0.9643       average accuracy = 0.7554)


Train: 100% 2000/2000 [01:51<00:00, 18.00 step/s, accuracy=0.88, loss=0.56, step=54000]
Valid: 100% 6944/6944 [00:14<00:00, 477.68 uttr/s, accuracy=0.70, loss=1.31]
Train:   0% 10/2000 [00:00<00:49, 39.96 step/s, accuracy=0.62, loss=1.43, step=54010]

average loss = 0.9457       average accuracy = 0.7635)


Train:  50% 1001/2000 [00:53<00:33, 29.68 step/s, accuracy=0.75, loss=0.89, step=55001]

average loss = 0.9368       average accuracy = 0.7641)


Train: 100% 2000/2000 [01:48<00:00, 18.41 step/s, accuracy=0.69, loss=1.13, step=56000]
Valid: 100% 6944/6944 [00:15<00:00, 446.93 uttr/s, accuracy=0.71, loss=1.28]
Train:   0% 7/2000 [00:00<02:51, 11.64 step/s, accuracy=0.72, loss=1.15, step=56007]

average loss = 0.9236       average accuracy = 0.7688)


Train:  50% 1001/2000 [00:55<00:45, 21.81 step/s, accuracy=0.75, loss=0.88, step=57001]

average loss = 0.9102       average accuracy = 0.7711)


Train: 100% 2000/2000 [01:51<00:00, 17.97 step/s, accuracy=0.78, loss=0.89, step=58000]
Valid: 100% 6944/6944 [00:15<00:00, 446.97 uttr/s, accuracy=0.71, loss=1.27]
Train:   0% 9/2000 [00:00<00:45, 43.46 step/s, accuracy=0.84, loss=0.78, step=58009]

average loss = 0.8870       average accuracy = 0.7760)


Train:  50% 1001/2000 [00:54<00:36, 27.09 step/s, accuracy=0.78, loss=0.96, step=59001]

average loss = 0.8849       average accuracy = 0.7783)


Train: 100% 2000/2000 [01:52<00:00, 17.70 step/s, accuracy=0.81, loss=0.94, step=6e+4]
Valid: 100% 6944/6944 [00:15<00:00, 439.46 uttr/s, accuracy=0.71, loss=1.28]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 3/2000 [00:00<02:23, 13.88 step/s, accuracy=0.69, loss=1.08, step=6e+4]

Step 60000, best model saved. (accuracy=0.7143)
average loss = 0.8898       average accuracy = 0.7779)


Train:  50% 1006/2000 [00:53<00:51, 19.12 step/s, accuracy=0.81, loss=0.71, step=61006]

average loss = 0.8851       average accuracy = 0.7808)


Train: 100% 2000/2000 [01:52<00:00, 17.81 step/s, accuracy=0.81, loss=0.55, step=62000]
Valid: 100% 6944/6944 [00:15<00:00, 454.48 uttr/s, accuracy=0.71, loss=1.25]
Train:   0% 5/2000 [00:00<03:11, 10.40 step/s, accuracy=0.91, loss=0.40, step=62005]

average loss = 0.8692       average accuracy = 0.7783)


Train:  50% 1001/2000 [00:54<00:44, 22.38 step/s, accuracy=0.81, loss=0.49, step=63001]

average loss = 0.8560       average accuracy = 0.7874)


Train: 100% 2000/2000 [01:56<00:00, 17.13 step/s, accuracy=0.81, loss=0.69, step=64000]
Valid: 100% 6944/6944 [00:15<00:00, 438.30 uttr/s, accuracy=0.72, loss=1.26]
Train:   0% 5/2000 [00:00<03:05, 10.75 step/s, accuracy=0.62, loss=1.49, step=64005]

average loss = 0.8508       average accuracy = 0.7837)


Train:  50% 1001/2000 [00:54<00:47, 21.24 step/s, accuracy=0.84, loss=0.82, step=65001]

average loss = 0.8676       average accuracy = 0.7851)


Train: 100% 2000/2000 [01:57<00:00, 17.04 step/s, accuracy=0.81, loss=0.53, step=66000]
Valid: 100% 6944/6944 [00:15<00:00, 460.25 uttr/s, accuracy=0.72, loss=1.23]
Train:   0% 10/2000 [00:00<00:41, 47.58 step/s, accuracy=0.75, loss=1.21, step=66010]

average loss = 0.8416       average accuracy = 0.7891)


Train:  50% 1003/2000 [00:56<01:03, 15.82 step/s, accuracy=0.66, loss=1.02, step=67003]

average loss = 0.8501       average accuracy = 0.7848)


Train: 100% 2000/2000 [01:59<00:00, 16.79 step/s, accuracy=0.84, loss=0.56, step=68000]
Valid: 100% 6944/6944 [00:15<00:00, 448.89 uttr/s, accuracy=0.72, loss=1.23]
Train:   0% 10/2000 [00:00<00:41, 48.28 step/s, accuracy=0.69, loss=1.43, step=68010]

average loss = 0.8512       average accuracy = 0.7877)


Train:  50% 1005/2000 [00:54<00:48, 20.35 step/s, accuracy=0.66, loss=1.41, step=69005]

average loss = 0.8326       average accuracy = 0.7912)


Train: 100% 2000/2000 [01:57<00:00, 16.97 step/s, accuracy=0.78, loss=1.02, step=7e+4]
Valid: 100% 6944/6944 [00:15<00:00, 447.15 uttr/s, accuracy=0.72, loss=1.25]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 0/2000 [00:00<?, ? step/s]


Step 70000, best model saved. (accuracy=0.7225)


######original : val acc = 68%
######d_model 80->128 , multihead 2->1
###### train acc = 80% , val acc = 73%

######TransformerEncoder layer : 1 -> 2 :

In [ ]:
import os
import json
import torch
from pathlib import Path
from torch.utils.data import Dataset


class InferenceDataset(Dataset):
  def __init__(self, data_dir):
    # using / to merge path
    testdata_path = Path(data_dir) / "testdata.json"
    metadata = json.load(testdata_path.open())
    self.data_dir = data_dir
    #
    self.data = metadata["utterances"]

  def __len__(self):
    return len(self.data)

  def __getitem__(self, index):
    utterance = self.data[index]
    feat_path = utterance["feature_path"]
    mel = torch.load(os.path.join(self.data_dir, feat_path))
    # compare : when we load training set, we use append to concate datas
    return feat_path, mel


def inference_collate_batch(batch):
  """Collate a batch of data."""
  feat_paths, mels = zip(*batch)

  return feat_paths, torch.stack(mels)

In [ ]:
import json
import csv
from pathlib import Path
from tqdm.notebook import tqdm

import torch
from torch.utils.data import DataLoader

def parse_args():
  """arguments"""
  config = {
    "data_dir": "./Dataset",
    "model_path": "./model.ckpt",
    "output_path": "./output.csv",
  }

  return config


def main(
  data_dir,
  model_path,
  output_path,
):
  """Main function."""
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  print(f"[Info]: Use {device} now!")

  mapping_path = Path(data_dir) / "mapping.json"
  mapping = json.load(mapping_path.open())

  dataset = InferenceDataset(data_dir)
  dataloader = DataLoader(
    dataset,
    batch_size=1,
    shuffle=False,
    drop_last=False,
    num_workers=8,
    collate_fn=inference_collate_batch,
  )
  print(f"[Info]: Finish loading data!",flush = True)

  speaker_num = len(mapping["id2speaker"])
  model = Classifier(n_spks=speaker_num).to(device)
  model.load_state_dict(torch.load(model_path))
  model.eval()
  print(f"[Info]: Finish creating model!",flush = True)

  results = [["Id", "Category"]]
  for feat_paths, mels in tqdm(dataloader):
    with torch.no_grad():
      mels = mels.to(device)
      outs = model(mels)
      preds = outs.argmax(1).cpu().numpy()
      # zip from two list to a sequence of tuples
      # preds is a numpy array whose elements are int numbers
      for feat_path, pred in zip(feat_paths, preds):
        # mapping["id2speaker"] = {"0": "id10001", ......'599':  'id11251'}])
        results.append([feat_path, mapping["id2speaker"][str(pred)]])

  with open(output_path, 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerows(results)


if __name__ == "__main__":
  main(**parse_args())


[Info]: Use cuda now!
[Info]: Finish loading data!
[Info]: Finish creating model!


  0%|          | 0/6000 [00:00<?, ?it/s]

In [ ]:
class MyLinearLayer(nn.Module):
    """ Custom Linear layer but mimics a standard linear layer """
    def __init__(self, size_in, size_out):
        super().__init__()
        self.size_in, self.size_out = size_in, size_out
        weights = torch.Tensor(size_out, size_in)
        self.weights = nn.Parameter(weights)  # nn.Parameter is a Tensor that's a module parameter.
        bias = torch.Tensor(size_out)
        self.bias = nn.Parameter(bias)

        # initialize weights and biases
        nn.init.kaiming_uniform_(self.weights, a=math.sqrt(5)) # weight init
        fan_in, _ = nn.init._calculate_fan_in_and_fan_out(self.weights)
        bound = 1 / math.sqrt(fan_in)
        nn.init.uniform_(self.bias, -bound, bound)  # bias init

    def forward(self, x):
        w_times_x= torch.mm(x, self.weights.t())
        return torch.add(w_times_x, self.bias)  # w times x + b
class MyLinear(nn.Module):
    def __init__(self, in_units, units):
        super().__init__()
        self.weight = nn.Parameter(torch.randn(in_units, units))
        self.bias = nn.Parameter(torch.randn(units,))

    def forward(self, X):
        linear = torch.matmul(X, self.weight.data) + self.bias.data
        return F.relu(linear)
class NoisyLinear(T.nn.Module):
  def __init__(self, n_in, n_out):
    super().__init__()
    self.n_in, self.n_out = n_in, n_out

    self.weights = T.nn.Parameter(T.zeros((n_out, n_in),
      dtype=T.float32))
    self.bias = T.nn.Parameter(T.tensor(n_out,
      dtype=T.float32))
    self.lo = 0.90; self.hi = 0.98  # noise

    lim = 0.01  # initialize weights and bias
    T.nn.init.uniform_(self.weights, -lim, +lim)
    T.nn.init.uniform_(self.bias, -lim, +lim)

  def forward(self, x):
    wx= T.mm(x, self.weights.t())
    rnd = (self.hi - self.lo) * T.rand(1) + self.lo
    return rnd * T.add(wx, self.bias)  # wts * x + bias